server.c

```diff
@@ -1292,6 +1297,96 @@ client_close(struct client *c)
 	client_close_ev(c->fd, 0, c);
 }
 
+static ssize_t
+read_cb(struct tls *ctx, void *buf, size_t buflen, void *cb_arg)
+{
+	struct client *c = cb_arg;
+
+	if (!c->should_buffer) {
+		// no buffer to cache into, read into libtls buffer
+		ssize_t ret = read(c->fd, buf, buflen);
+		if (-1 == ret && errno == EWOULDBLOCK)
+			ret = TLS_WANT_POLLIN;
+		
+		return ret;
+	}
+
+	if (c->buf.has_tail) {
+		// we have leftover data from a previous call to read_cb
+		size_t left = BUFLAYER_MAX - c->buf.read_pos;
+		size_t copy_len = MINIMUM(buflen, left);
+		memcpy(buf, c->buf.data + c->buf.read_pos, copy_len);
+
+		c->buf.read_pos += copy_len;
+
+		if (left == copy_len) {
+			// memset(buflayer, 0, BUFLAYER_MAX);
+			c->should_buffer = 0;
+			c->buf.has_tail = 0;
+		}
+		
+		return copy_len;
+	}
+
+	// buffer layer exists, we expect proxy protocol
+	errno = 0;
```

Comment:
nit: setting errno is not needed.  read will set it if it fails, so no need to explicitly zero it.